arch/xtensa/src/common/xtensa_initialstate.c

```diff
@@ -100,6 +124,71 @@ void up_initial_state(struct tcb_s *tcb)
   xcp->regs[REG_A1] = (uint32_t)tcb->stack_base_ptr + /* Physical top of stack frame    */
                                 tcb->adj_stack_size;
 
+  /* Each task access the TLS variables using the THREADPTR register plus an
+   * offset to obtain the address of the variable.
+   *
+   * TLS layout at link-time, where 0xNNN is the offset that the linker
+   * calculates to a particular TLS variable.
+   *
+   * LOW ADDRESS
+   *         |---------------------------|   Linker Symbols
+   *         | Section                   |   --------------
+   *         | .flash.rodata             |
+   *      0x0|---------------------------| <- _rodata_reserved_start
+   *       ^ | Other Data                |
+   *       | |---------------------------| <- _thread_local_start
+   *       | | .tbss                     | ^
+   *       V |                           | |
+   *   0xNNN | int example;              | | tls_area_size
+   *         |                           | |
+   *         | .tdata                    | V
+   *         |---------------------------| <- _thread_local_end
+   *         | Other data                |
+   *         | ...                       |
+   *         |---------------------------|
+   * HIGH ADDRESS
+   *
+   * Calculate the THREADPTR register's initialization value based on the
+   * link-time offset and the TLS area allocated on the stack.
+   *
+   * HIGH ADDRESS
+   *         |---------------------------| <- top of the Stack
+   *         |                           |
+   *         |                           |
+   *         |                           |
+   *         |                           |
+   *         |---------------------------|
+   *         | .tdata (*)                |
+   *       ^ | int example;              |
+   *       | |                           |
+   *       | | .tbss (*)                 |
+   *       | |---------------------------| <- start of TLS area
+   * 0xNNN | |                           | ^
+   *       | |                           | |
+   *       | |           ...             | | (_thread_local_start -
+   *       | |                           | |   _rodata_reserved_start)
+   *       | |                           | |  + align_up(TCB_SIZE,
+   *       | |                           | |             tls_section_align)
+   *       | |                           | |
+   *       | |                           | V
+   *       V |                           | <- threadptr register's value
+   *
+   * LOW ADDRESS
+   *
+   * Note: Xtensa is slightly different compared to the RISC-V port as there
+   * is an implicit aligned TCB_SIZE added to the offset.
+   * (search for 'tpoff' in elf32-xtensa.c in BFD):
```

Comment:
this one? https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=bfd/elf32-xtensa.c;h=f078cbde7146675fd2ed82eb5102ae2596c20775;hb=HEAD#l1830

i guess it's handy to have an url here.

Reply:
Exactly. Thanks for pointing this out. I'll update the sources to include it ;)